### ISSUE TITLE
Deprecate or remove SHA1 support

### ISSUE THREAD

**Issue by stianst** [keycloak/keycloak#32897]
### Description

SHA1 should no longer be used anywhere, by anyone. We should deprecate or remove all use of SHA1.

If we do this in a minor release we will probably need to add a feature guard of sorts around this, and make it possible to disable SHA1 support, to then remove it later in a major.

However, it may not even work to use SHA1 at this point, as it may be disallowed in Java 17. This would need some further investigations.

**Comment by abstractj** [comment]
@stianst although I agree that SHA-1 should be deprecated/removed, [Java 17 will still allow its use](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/security/MessageDigest.html). However, only Jarsigner will be affected—jars signed with SHA-1 will be treated as unsigned because the weak algorithm is now disabled by the security property.

There's one caveat about FIPS and @mposolda please keep me honest here. Today we are at FIPS 140-2 and to remain compliant we must support SHA-1, although on FIPS 140-3 (_not supported by us_) it was deprecated.

Now planning the deprecation, or removal, I would like to create sub-tasks for the items below if you both agree:

### **1. `adapters/saml`**

- **Schemas**: Multiple versions of `keycloak_saml_adapter_1_*.xsd` and `wildfly-keycloak-saml_1_*.xsd` reference `RSA_SHA1` and `DSA_SHA1`.  
    Action: Remove or mark these enums as deprecated.
- **Test XML Configs**: `keycloak-saml-1.4.xml` uses `signatureAlgorithm="DSA_SHA1"`.  
    Action: Update to stronger algorithms (e.g., `RSA_SHA256`).

### **2. `core`**

- **JWK.java / RSAPublicJWK.java**: Uses `SHA1` to generate x.509 thumbprints.  
    Action: Migrate to SHA-256 (`x5t#S256`) and deprecate `x5t`.
- **Tests**: `JWKTest.java` and `PemUtilsTest.java` assert and test SHA1 thumbprints.  
    Action: Replace with SHA-256 equivalents.

### **3. `crypto` (default, elytron, fips)**

- **SHA1PRNG**: Used across providers and tests.  
    Action: Replace with more secure alternatives (e.g., `DRBG`, `NativePRNG`).
- **DigestCalculator**: References `HASH_SHA1`.  
    Action: Remove or replace with `HASH_SHA256` or higher.

### **4. `docs`**

- Multiple files note SHA-1 deprecation (e.g., `otp-policies.adoc`, `clients/saml/*.adoc`).  
    Action: Update examples and guidance to highlight SHA256/SHA512 only.

### **5. `federation/sssd`**

- **SASL.java** and config files: `DBUS_COOKIE_SHA1` references.  
    Action: Investigate feasibility of removing or phasing out support.

### **6. `js/apps/admin-ui`**

- **OTP Algorithm Lists** and **SAML Signature Options** include SHA1.  
    Action: Remove from UI dropdowns and update i18n messages.

### **7. `model/jpa`**

- **Liquibase scripts**: Default values include `HmacSHA1`.  
    Action: Change default to `HmacSHA256` or `HmacSHA512`.

### **8. `operator`**

- **Test Realms**: Use `otpPolicyAlgorithm: HmacSHA1`.  
    Action: Replace test config with stronger algorithms.

### **9. `saml-core-api` / `saml-core`**

- **Constants and Enums**: Define and use `RSA_SHA1`, `DSA_SHA1`.  
    Action: Remove or deprecate these constants and update mappings.
- **Default Algorithms**: Default to SHA1 in places like `BaseSAML2BindingBuilder`.  
    Action: Set default to `RSA_SHA256`.
- **Tests and Metadata**: Use SHA1 in XML metadata.  
    Action: Replace metadata and update expected test values.

### **10. `server-spi-private`**

- **PBKDF2**: Uses `PBKDF2WithHmacSHA1`.  
    Action: Use `PBKDF2WithHmacSHA256` by default.
- **SAML Utils**: `MessageDigest.getInstance("SHA-1")` and exceptions referencing SHA-1.  
    Action: Migrate to SHA-256 and update error messages.
- **Tests**: OTP and password hashing tests rely on SHA1.  
    Action: Replace with stronger algorithms.

### **11. `server-spi`**

- **OTP Policy**: Default algorithm is `HmacSHA1`.  
    Action: Update to `HmacSHA256` or allow user selection only from strong algorithms.
- **HmacOTP.java**: Hardcoded default = HMAC_SHA1.  
    Action: Change to HMAC_SHA256 and document change.

### **12. `services`**

- **DPoPUtil** and `SAML Artifact Resolver`: Use SHA1 for hashing and secure random.  
    Action: Switch to SHA-256 and more secure PRNGs.
- **OIDC Identity Provider**: Uses x5t thumbprint (`SHA1`).  
    Action: Add x5t#S256 and prefer it in JWS headers.
### **13. `testsuite`**

- **Tests and Resources**: Widespread use of HmacSHA1 in OTP, hashing, SAML metadata, and Kerberos test configs.  
    Action: Refactor all test realms, test cases, and mock data to use stronger algorithms.

**Comment by mposolda** [comment]
Moving to 27, so we can remove SHA1 without too much care about backwards compatibility.

Created the task for the deprecation: https://github.com/keycloak/keycloak/issues/40706 . We can address that one in Keycloak 26.4, so we can handle this one in 27

**Comment by dalbani** [comment]
Just to be sure, does "removing SHA1 support" mean that all the existing passwords with algorithm `pbkdf2` and OTP credentials with algorithm `HmacSHA1` will suddenly stop working in Keycloak 27?

**Comment by abstractj** [comment]
@dalbani from my understanding @mposolda plans to deprecate first, before removing.

**Comment by dalbani** [comment]
That's really the removal that I'm most worried about 😄 
As it would potentially prevent existing users with SHA1 based credentials (password, TOTP) from logging in, if I correctly understand the consequences of removing SHA1 support altogether in Keycloak.

**Comment by marconett** [comment]
Just a heads up to the Keycloak team from someone currently implementing a TOTP system using Keycloak:

Removing the ability to use HMAC-SHA1 as an "OTP hash algorithm" in the OTP policy would be a huge breaking change.  

Upgrading a Keycloak instance to a version without HMAC-SHA1 support would lock out all users whose authenticator relies on this algorithm. This would effectively result in one support case for each TOTP user that needs to be manually rectified, as there's no way to "upgrade" their existing HMAC-SHA1 authenticator.

Given that HMAC-SHA1 is currently the default OTP policy in Keycloak (and in most authenticator apps, for that matter), a large amount of Keycloak instances would be confronted with this issue.

Even deprecating the algorithm first - and thereby giving the Keycloak instance admins a "heads up" - is a huge headache, as this still requires the end-user to get active and repeat the authenticator setup process to upgrade to (for example) HMAC-SHA256.

As a side note, HMAC-SHA1 is not the same as SHA1 and is not vulnerable to the same attack vector (see discussion here: https://crypto.stackexchange.com/questions/26510/why-is-hmac-sha1-still-considered-secure)

**Comment by kerki** [comment]
@stianst I have some concerns about this proposed change. Moving away from SHA-1 for OTP generation would effectively lock out all Microsoft Authenticator users, as that app currently only supports HMAC-SHA1 for TOTP.

While SHA-1 is considered weak for digital signatures and long-term integrity protection, this does not apply to TOTP. In the OTP use case, SHA-1 is only used in an HMAC over a short-lived counter value, which makes practical attacks infeasible.

Switching the default algorithm could therefore cause widespread compatibility issues without providing a real security benefit in this context. A better approach might be to keep HMAC-SHA1 as the default while allowing stronger algorithms as an option for clients that support them.

**Comment by mmrhw** [comment]
@stianst 
I agree with @marconett and @kerki . My company is also currently planning a project with Keycloak and we are also of the opinion that HMAC-SHA1 is not a problem in connection with OTP. As Microsoft Authenticator is one of the two most popular authenticator apps, Keycloak not supporting this app would have a negative impact on our customer experience and therefore on our business. We would absolutely support this change if the security concerns were widespread. However, this does not seem to be the case - we would therefore ask you to reconsider this change.

We'd be happy if you could share your thoughts on the objections with us.

**Comment by stianst** [comment]
To everyone that is worried about us suddenly removing SHA-1 support causing issues with older clients, otp, password hashing, etc. that is obviously not how we will do things around this; we may remove some support, deprecate others, or simply provide recommendations. This issue is mostly just a placeholder for now; but we will break this up into smaller bits and pieces.
